### DISCUSSION TITLE
How to Build and Release packages with OS/Platform variants

### DISCUSSION THREAD

**Discussion by kalmarmeter** [AcademySoftwareFoundation/rez#1547] [Q&A] [answered]
Hello everyone!

In the last couple weeks I have been trying to set rez up for my already existing Maya plugins.
These plugins have to be compiled and installed for both windows and linux versions.

I ran into an issue when handling the different platforms with Variants.
I defined 2 Variants, one for each platform (and maya version as an added variant, so far having one single value):
`variants = [
    ['maya-2022', 'platform-windows'],
    ['maya-2022', 'platform-linux'],
]`

However, when I now want to call `rez build --install` or `rez release`, these calls will fail, since one of the variants will conflict with my workstation's platform and cannot be built locally.

I could do `rez build --install --variants 0` on Windows, and `--variants 1` on Linux, however this is seems tedious, error-prone and still does not solve the issue when releasing. Also, if my package's variant list changes, all these variant indices change too.

Am I missing something obvious here?
How would you handle multiplatform package builds, installs and releases?

Thank you for your help!
Peter

**Comment by JeanChristopheMorinPerso** [comment, marked as answer]
One way to solve this is to use the `expand_requires` function like this:
```python

@early
def variants():
    import platform
    from rez.package_py_utils import expand_requires

    if platform.system() not in ["Linux", "Windows"]:
        # I'm not 100% sure what exception we should raise though.
        raise RuntimeError("asdasd")

    return [["maya-2022"] + expand_requires("platform-**")]
```

See https://rez.readthedocs.io/en/stable/api/rez.package_py_utils.html#rez.package_py_utils.expand_requires and https://rez.readthedocs.io/en/stable/package_definition.html#early-binding-functions.

**Reply by kalmarmeter** [reply]
Thank you, this solved my problems. Some small corrections I had to do to the code:
- The early decorator has to be written with parantheses, likethis: `early()`
- Expand requires did not accept the * wildcard, so I just used `expand_requires('platform')`

**Just for clarification for future programmers:**
Releasing from different platforms is multiple, separate rez release calls from each machine. However rez allows releasing different variants of the same version of a package one at a time.
So a CICD setup may do rez release on each machine and each would only release their given platform variants.
